services/motion/builtin/state/state.go

```diff
@@ -0,0 +1,619 @@
+							return
+						}
+						e.testConfig.ExecutionResponseChan <- struct{}{}
+						return
+					}
+
+					e.logger.Debugf("plan succeeded %#v\n", lastPlan)
+					if err := e.notifyStatePlanSucceeded(lastPlan, time.Now()); err != nil {
+						e.logger.Error(err.Error())
+						e.testConfig.ExecutionResponseChan <- struct{}{}
+						return
+					}
+					e.testConfig.ExecutionResponseChan <- struct{}{}
+					return
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (e *execution) toStateExecution() stateExecution {
+	return stateExecution{
+		id:            e.id,
+		componentName: e.componentName,
+		waitGroup:     e.waitGroup,
+		cancelFunc:    e.cancelFunc,
+	}
+}
+
+// NOTE: We hold the lock for both updateStateNewExecution & updateStateNewPlan to ensure no readers
+// are able to see a state where the execution exists but does not have a plan with a status.
+func (e *execution) notifyStateNewExecution(execution stateExecution, plan motion.Plan, time time.Time) error {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+
+	if err := e.state.updateStateNewExecution(execution); err != nil {
+		return err
+	}
+
+	msg := newPlanMsg{
+		plan:       plan,
+		planStatus: motion.PlanStatus{State: motion.PlanStateInProgress, Timestamp: time},
+	}
+	if err := e.state.updateStateNewPlan(msg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *execution) notifyStateNewPlan(plan motion.Plan, time time.Time) error {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	return e.state.updateStateNewPlan(newPlanMsg{
+		plan:       plan,
+		planStatus: motion.PlanStatus{State: motion.PlanStateInProgress, Timestamp: time},
+	})
+}
+
+func (e *execution) notifyStatePlanFailed(plan motion.Plan, reason string, time time.Time) error {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	return e.state.updateStateStatusUpdate(stateUpdateMsg{
+		componentName: e.componentName,
+		executionID:   e.id,
+		planID:        plan.ID,
+		planStatus:    motion.PlanStatus{State: motion.PlanStateFailed, Timestamp: time, Reason: &reason},
+	})
+}
+
+func (e *execution) notifyStatePlanSucceeded(plan motion.Plan, time time.Time) error {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	return e.state.updateStateStatusUpdate(stateUpdateMsg{
+		componentName: e.componentName,
+		executionID:   e.id,
+		planID:        plan.ID,
+		planStatus:    motion.PlanStatus{State: motion.PlanStateSucceeded, Timestamp: time},
+	})
+}
+
+// a stateExecution is the struct held in the state that
+// holds the history of plans & plan status updates an
+// execution has exprienced & the waitGroup & cancelFunc
+// required to shut down an execution's goroutine.
+type stateExecution struct {
+	id            motion.ExecutionID
+	componentName resource.Name
+	waitGroup     *sync.WaitGroup
+	cancelFunc    context.CancelFunc
+	history       []motion.PlanWithStatus
+}
+
+func (e *stateExecution) stop() {
+	e.cancelFunc()
+	e.waitGroup.Wait()
+}
+
+var replanReason = "replanning"
```

Comment:
Why is this here?

Reply:
b/c this is the string returned as the reason when the reason why the Plan failed is due to replanning